### PR TITLE
[JENKINS-34755] fix read of 'SystemProperties' values before init of the context

### PR DESCRIPTION
**Pull request by evernat**
Not tested. But I am quite sure that this PR fixes read of 'SystemProperties' values before init of the context.
See https://github.com/jenkinsci/jenkins/pull/2346#issuecomment-219216327

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
will it work with the private constructor?
Need to RTFM a bit

**Comment by oleg-nenashev** [comment]
tests need to be adjusted

**Comment by evernat** [comment]
Good catch: constructor needs to be public.
I did not see that test directory next to core. It's fixed now.

**Comment by oleg-nenashev** [comment]
Merge conflict :(

File: test/src/test/java/jenkins/util/SystemPropertiesTest.java [hunk 1]
```diff
@@ -43,7 +44,7 @@
     
     @Before
     public void setUp() {
-        SystemProperties.initialize(j.jenkins.servletContext);
+        new SystemProperties().contextInitialized(new ServletContextEvent(j.jenkins.servletContext));
```
Comment (oleg-nenashev) [inline test/src/test/java/jenkins/util/SystemPropertiesTest.java]: Does not seem to be a perfect interface to the static method, but OK

**Comment by oleg-nenashev** [comment]
@evernat 
Thanks for addressing the comments! I'll try it manually today

**Comment by oleg-nenashev** [comment]
With this change I'm getting the following message on startup:

```
May 15, 2016 2:16:45 PM org.eclipse.jetty.util.log.JavaUtilLog warn
WARNING: Could not instantiate listener jenkins.util.SystemProperties
java.lang.IllegalAccessException: Class org.eclipse.jetty.server.handler.ContextHandler$Context can not access a member of class jenkins.util.SystemProperties with modifiers "private"
	at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:102)
	at java.lang.Class.newInstance(Class.java:436)
	at org.eclipse.jetty.server.handler.ContextHandler$Context.createInstance(ContextHandler.java:2372)
	at org.eclipse.jetty.webapp.StandardDescriptorProcessor.newListenerInstance(StandardDescriptorProcessor.java:2014)
	at org.eclipse.jetty.webapp.StandardDescriptorProcessor.visitListener(StandardDescriptorProcessor.java:1957)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.eclipse.jetty.webapp.IterativeDescriptorProcessor.visit(IterativeDescriptorProcessor.java:85)
	at org.eclipse.jetty.webapp.IterativeDescriptorProcessor.process(IterativeDescriptorProcessor.java:72)
	at org.eclipse.jetty.webapp.MetaData.resolve(MetaData.java:408)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1340)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:741)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:505)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:132)
	at org.eclipse.jetty.server.Server.start(Server.java:387)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:114)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
	at org.eclipse.jetty.server.Server.doStart(Server.java:354)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at winstone.Launcher.<init>(Launcher.java:152)
	at winstone.Launcher.main(Launcher.java:352)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at Main._main(Main.java:246)
	at Main.main(Main.java:91)
```

**Comment by oleg-nenashev** [comment]
After proper build it works like a charm 
👍

**Comment by oleg-nenashev** [comment]
Timeout on CI :(

**Comment by batmat** [comment]
`docker run -ti -p 8080:8080 -e ID=2347 batmat/jenkins-pr-tester` works fine and code LGTM.
:+1:

**Comment by oleg-nenashev** [comment]
Thanks. My custom WAR file uses several properties coming from the context, and all of them work correctly after the merge. Ready to go.

I'd ignore the CI run, because something weird is going on with its performance

**Comment by oleg-nenashev** [comment]
Thank a lot @evernat! If you have any use-cases for this Engine, it would be great to know about them

**Pull request merged**